File: backend-python/app/modules/system/service/dept_service.py
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.common.errors import BusinessError, bad_request, not_found
from app.common.id import new_id
from app.modules.system.dto.dept_request import DeptSaveRequest
from app.modules.system.entity import SysDept, SysUser


class DeptService:
    def __init__(self, db: AsyncSession) -> None:
        self.db = db

    async def tree(self) -> list[dict]:
        rows = list(
            (
                await self.db.scalars(
                    select(SysDept)
                    .where(SysDept.deleted == 0)
                    .order_by(SysDept.sort, SysDept.id)
                )
            ).all()
        )
        nodes = {row.id: self._vo(row) for row in rows}
        roots: list[dict] = []
        for row in rows:
            parent = nodes.get(row.parent_id)
            if parent is None:
                roots.append(nodes[row.id])
            else:
                parent["children"].append(nodes[row.id])
        return roots

    async def detail(self, dept_id: int) -> dict:
        return self._vo(await self._find(dept_id))

    async def save(self, dept_id: int | None, request: DeptSaveRequest) -> str | None:
        parent_id = request.parent_id or 0
        await self._assert_valid_parent(dept_id, parent_id)
        await self._assert_leader(request.leader_user_id)
        duplicate = select(SysDept.id).where(
            SysDept.dept_code == request.dept_code.strip(),
            SysDept.deleted == 0,
        )
        if dept_id is not None:
            duplicate = duplicate.where(SysDept.id != dept_id)
        if await self.db.scalar(duplicate) is not None:
            raise BusinessError(409000, "deptCode already exists")
        values = {
            "parent_id": parent_id,
            "dept_name": request.dept_name.strip(),
            "dept_code": request.dept_code.strip(),
            "leader_user_id": request.leader_user_id,
            "sort": 0 if request.sort is None else request.sort,
            "status": 1 if request.status is None else request.status,
        }
        if dept_id is None:
            row = SysDept(id=new_id(), deleted=0, **values)
            self.db.add(row)
            await self.db.commit()
            return str(row.id)
        row = await self._find(dept_id)
        for key, value in values.items():
            setattr(row, key, value)
        await self.db.commit()
        return None

    async def delete(self, dept_id: int) -> None:
        row = await self._find(dept_id)
        child_count = await self.db.scalar(
            select(func.count()).select_from(SysDept).where(
                SysDept.parent_id == dept_id,
                SysDept.deleted == 0,
            )
        )
        if child_count:
            raise BusinessError(409000, "department contains child departments")
        user_count = await self.db.scalar(
            select(func.count()).select_from(SysUser).where(
                SysUser.dept_id == dept_id,
                SysUser.deleted == 0,
            )
        )
        if user_count:
            raise BusinessError(409000, "department contains users")
        row.deleted = 1
        await self.db.commit()

    async def update_status(self, dept_id: int, status: int | None) -> None:
        row = await self._find(dept_id)
        row.status = 1 if status is None else status
        await self.db.commit()

    async def _assert_valid_parent(self, dept_id: int | None, parent_id: int) -> None:
        if parent_id == 0:
            return
        await self._find(parent_id)
        if dept_id is None:
            return
        if parent_id == dept_id:
            raise bad_request("parent department cannot be self or descendant")
        rows = (await self.db.execute(select(SysDept.id, SysDept.parent_id).where(SysDept.deleted == 0))).all()
        if parent_id in _descendants(dept_id, rows):
            raise bad_request("parent department cannot be self or descendant")

    async def _assert_leader(self, user_id: int | None) -> None:
        if user_id is None:
            return
        exists = await self.db.scalar(
            select(SysUser.id).where(SysUser.id == user_id, SysUser.deleted == 0)
        )
        if exists is None:
            raise bad_request("leaderUserId is invalid")

    async def _find(self, dept_id: int) -> SysDept:
        row = await self.db.scalar(select(SysDept).where(SysDept.id == dept_id, SysDept.deleted == 0))
        if row is None:
            raise not_found()
        return row

    @staticmethod
    def _vo(row: SysDept) -> dict:
        return {
            "id": str(row.id),
            "parentId": str(row.parent_id),
            "deptName": row.dept_name,
            "deptCode": row.dept_code,
            "leaderUserId": str(row.leader_user_id) if row.leader_user_id is not None else None,
            "sort": row.sort,
            "status": row.status,
            "createdAt": row.created_at,
            "updatedAt": row.updated_at,
            "children": [],
        }


def _descendants(parent_id: int, rows: list[tuple[int, int]]) -> set[int]:
    children: dict[int, list[int]] = {}
    for child_id, current_parent_id in rows:
        children.setdefault(current_parent_id, []).append(child_id)
    result: set[int] = set()
    stack = list(children.get(parent_id, []))
    while stack:
        current = stack.pop()
        if current in result:
            continue
        result.add(current)
        stack.extend(children.get(current, []))
    return result

File: backend-rust/src/modules/system/service/dept_service.rs
use crate::common::{AppError, I64String, next_id};
use crate::modules::system::dto::dept_request::DeptSaveRequest;
use crate::modules::system::entity::sys_dept::SysDept;
use crate::modules::system::vo::dept_tree_vo::DeptTreeVo;
use rbatis::RBatis;
use serde::Deserialize;
use std::collections::{BTreeMap, BTreeSet};
use std::sync::Arc;

#[derive(Debug, Deserialize)]
struct CountRow {
    total: i64,
}

pub async fn list(database: Option<&Arc<RBatis>>) -> Result<Vec<DeptTreeVo>, AppError> {
    let database = require_database(database)?;
    let rows = all_depts(database).await?;
    Ok(build_tree(rows))
}

pub async fn detail(database: Option<&Arc<RBatis>>, id: i64) -> Result<SysDept, AppError> {
    let database = require_database(database)?;
    detail_with_database(database, id).await
}

pub async fn create(
    database: Option<&Arc<RBatis>>,
    request: DeptSaveRequest,
) -> Result<I64String, AppError> {
    let database = require_database(database)?;
    validate_request(&request)?;
    let parent_id = request.parent_id.unwrap_or(0);
    assert_existing_parent(database, parent_id).await?;
    assert_unique_dept_code(database, None, &request.dept_code).await?;
    assert_existing_leader(database, request.leader_user_id).await?;

    let id = next_id();
    database
        .exec(
            "insert into sys_dept (id, parent_id, dept_name, dept_code, leader_user_id, sort, status, deleted) values (?, ?, ?, ?, ?, ?, ?, 0)",
            vec![
                rbs::value!(id),
                rbs::value!(parent_id),
                rbs::value!(request.dept_name),
                rbs::value!(request.dept_code),
                rbs::value!(request.leader_user_id),
                rbs::value!(request.sort.unwrap_or(0)),
                rbs::value!(request.status.unwrap_or(1)),
            ],
        )
        .await
        .map_err(map_database_error)?;
    Ok(I64String(id))
}

pub async fn update(
    database: Option<&Arc<RBatis>>,
    id: i64,
    request: DeptSaveRequest,
) -> Result<(), AppError> {
    let database = require_database(database)?;
    validate_request(&request)?;
    detail_with_database(database, id).await?;
    let parent_id = request.parent_id.unwrap_or(0);
    assert_valid_parent(database, id, parent_id).await?;
    assert_unique_dept_code(database, Some(id), &request.dept_code).await?;
    assert_existing_leader(database, request.leader_user_id).await?;

    database
        .exec(
            "update sys_dept set parent_id = ?, dept_name = ?, dept_code = ?, leader_user_id = ?, sort = ?, status = ? where id = ? and deleted = 0",
            vec![
                rbs::value!(parent_id),
                rbs::value!(request.dept_name),
                rbs::value!(request.dept_code),
                rbs::value!(request.leader_user_id),
                rbs::value!(request.sort.unwrap_or(0)),
                rbs::value!(request.status.unwrap_or(1)),
                rbs::value!(id),
            ],
        )
        .await
        .map_err(map_database_error)?;
    Ok(())
}

pub async fn delete(database: Option<&Arc<RBatis>>, id: i64) -> Result<(), AppError> {
    let database = require_database(database)?;
    detail_with_database(database, id).await?;
    let child_count = count(
        database,
        "select count(*) as total from sys_dept where parent_id = ? and deleted = 0",
        vec![rbs::value!(id)],
    )
    .await?;
    if child_count > 0 {
        return Err(AppError::bad_request("存在子部门，不能删除"));
    }
    let user_count = count(
        database,
        "select count(*) as total from sys_user where dept_id = ? and deleted = 0",
        vec![rbs::value!(id)],
    )
    .await?;
    if user_count > 0 {
        return Err(AppError::bad_request("部门下存在用户，不能删除"));
    }
    database
        .exec(
            "update sys_dept set deleted = 1 where id = ? and deleted = 0",
            vec![rbs::value!(id)],
        )
        .await
        .map_err(map_database_error)?;
    Ok(())
}

pub async fn update_status(
    database: Option<&Arc<RBatis>>,
    id: i64,
    status: i32,
) -> Result<(), AppError> {
    let database = require_database(database)?;
    detail_with_database(database, id).await?;
    database
        .exec(
            "update sys_dept set status = ? where id = ? and deleted = 0",
            vec![rbs::value!(status), rbs::value!(id)],
        )
        .await
        .map_err(map_database_error)?;
    Ok(())
}

async fn all_depts(database: &RBatis) -> Result<Vec<SysDept>, AppError> {
    database
        .exec_decode(
            "select id, parent_id, dept_name, dept_code, leader_user_id, sort, status, deleted, created_at, updated_at from sys_dept where deleted = 0 order by sort asc, id asc",
            vec![],
        )
        .await
        .map_err(map_database_error)
}

async fn detail_with_database(database: &RBatis, id: i64) -> Result<SysDept, AppError> {
    let rows: Vec<SysDept> = database
        .exec_decode(
            "select id, parent_id, dept_name, dept_code, leader_user_id, sort, status, deleted, created_at, updated_at from sys_dept where id = ? and deleted = 0",
            vec![rbs::value!(id)],
        )
        .await
        .map_err(map_database_error)?;
    rows.into_iter()
        .next()
        .ok_or_else(|| AppError::not_found("资源不存在"))
}

async fn assert_valid_parent(database: &RBatis, id: i64, parent_id: i64) -> Result<(), AppError> {
    if parent_id == 0 {
        return Ok(());
    }
    if parent_id == id
        || descendant_dept_ids(database, id)
            .await?
            .contains(&parent_id)
    {
        return Err(AppError::bad_request("上级部门不能选择自己或子部门"));
    }
    assert_existing_parent(database, parent_id).await
}

async fn assert_existing_parent(database: &RBatis, parent_id: i64) -> Result<(), AppError> {
    if parent_id == 0 {
        return Ok(());
    }
    detail_with_database(database, parent_id).await?;
    Ok(())
}

async fn assert_unique_dept_code(
    database: &RBatis,
    id: Option<i64>,
    dept_code: &str,
) -> Result<(), AppError> {
    let total = match id {
        Some(id) => {
            count(
                database,
                "select count(*) as total from sys_dept where dept_code = ? and id <> ? and deleted = 0",
                vec![rbs::value!(dept_code), rbs::value!(id)],
            )
            .await?
        }
        None => {
            count(
                database,
                "select count(*) as total from sys_dept where dept_code = ? and deleted = 0",
                vec![rbs::value!(dept_code)],
            )
            .await?
        }
    };
    if total > 0 {
        return Err(AppError::bad_request("部门编码已存在"));
    }
    Ok(())
}

async fn assert_existing_leader(
    database: &RBatis,
    leader_user_id: Option<i64>,
) -> Result<(), AppError> {
    let Some(leader_user_id) = leader_user_id else {
        return Ok(());
    };
    let total = count(
        database,
        "select count(*) as total from sys_user where id = ? and deleted = 0",
        vec![rbs::value!(leader_user_id)],
    )
    .await?;
    if total == 0 {
        return Err(AppError::bad_request("负责人用户不存在"));
    }
    Ok(())
}

async fn descendant_dept_ids(database: &RBatis, id: i64) -> Result<BTreeSet<i64>, AppError> {
    let rows = all_depts(database).await?;
    let mut children_by_parent: BTreeMap<i64, Vec<i64>> = BTreeMap::new();
    for row in rows {
        children_by_parent
            .entry(row.parent_id.value())
            .or_default()
            .push(row.id.value());
    }
    let mut result = BTreeSet::new();
    collect_descendants(id, &children_by_parent, &mut result);
    Ok(result)
}

fn collect_descendants(
    id: i64,
    children_by_parent: &BTreeMap<i64, Vec<i64>>,
    result: &mut BTreeSet<i64>,
) {
    if let Some(children) = children_by_parent.get(&id) {
        for child_id in children {
            if result.insert(*child_id) {
                collect_descendants(*child_id, children_by_parent, result);
            }
        }
    }
}

async fn count(database: &RBatis, sql: &str, args: Vec<rbs::Value>) -> Result<i64, AppError> {
    let counts: Vec<CountRow> = database
        .exec_decode(sql, args)
        .await
        .map_err(map_database_error)?;
    Ok(counts.first().map(|row| row.total).unwrap_or(0))
}

fn build_tree(rows: Vec<SysDept>) -> Vec<DeptTreeVo> {
    let ids: BTreeSet<i64> = rows.iter().map(|row| row.id.value()).collect();
    let mut children_by_parent: BTreeMap<i64, Vec<SysDept>> = BTreeMap::new();
    for row in rows {
        let parent_id = row.parent_id.value();
        let key = if parent_id == 0 || !ids.contains(&parent_id) {
            0
        } else {
            parent_id
        };
        children_by_parent.entry(key).or_default().push(row);
    }
    build_children(0, &mut children_by_parent)
}

fn build_children(
    parent_id: i64,
    children_by_parent: &mut BTreeMap<i64, Vec<SysDept>>,
) -> Vec<DeptTreeVo> {
    let rows = children_by_parent.remove(&parent_id).unwrap_or_default();
    rows.into_iter()
        .map(|row| {
            let id = row.id.value();
            DeptTreeVo {
                id: row.id,
                parent_id: row.parent_id,
                dept_name: row.dept_name,
                dept_code: row.dept_code,
                leader_user_id: row.leader_user_id,
                sort: row.sort,
                status: row.status,
                created_at: row.created_at,
                updated_at: row.updated_at,
                children: build_children(id, children_by_parent),
            }
        })
        .collect()
}

fn require_database(database: Option<&Arc<RBatis>>) -> Result<&RBatis, AppError> {
    database
        .map(|database| database.as_ref())
        .ok_or_else(|| AppError::system("Rbatis database is not configured"))
}

fn validate_request(request: &DeptSaveRequest) -> Result<(), AppError> {
    require_text(&request.dept_name, "deptName")?;
    require_text(&request.dept_code, "deptCode")?;
    validate_length(&request.dept_name, "deptName", 64)?;
    validate_length(&request.dept_code, "deptCode", 64)?;
    Ok(())
}

fn require_text(value: &str, field: &str) -> Result<(), AppError> {
    if value.trim().is_empty() {
        return Err(AppError::bad_request(format!("{field} is required")));
    }
    Ok(())
}

fn validate_length(value: &str, field: &str, max: usize) -> Result<(), AppError> {
    if value.chars().count() > max {
        return Err(AppError::bad_request(format!(
            "{field} length must be <= {max}"
        )));
    }
    Ok(())
}

fn map_database_error(err: rbatis::Error) -> AppError {
    let message = err.to_string();
    if message.to_ascii_lowercase().contains("duplicate") {
        AppError::conflict("数据冲突")
    } else {
        AppError::system(message)
    }
}
